optimization.ts

```diff
@@ -0,0 +1,433 @@
+    var functionOptimized: boolean = false;
+    //Write code to optimize functions here
+    var newBody: Array<BasicBlock<any>> = func.body.map(b => {
+        var tempBlockEnv: Env = duplicateEnv(inEnvMapping.get(b.label));
+        var [optimizedBlock, blockOptimized]: [BasicBlock<any>, boolean] = optimizeBlock(b, tempBlockEnv);
+        if (!functionOptimized && blockOptimized) functionOptimized = true;
+        return optimizedBlock;
+    });
+
+    if (functionOptimized) return optimizeFunction({...func, body: newBody})
+
+    return {...func, body: newBody};
+}
+
+export function optimizeClass(c: Class<any>): Class<any>{
+    var optimizedMethods: Array<FunDef<any>> = c.methods.map(m => {
+        return optimizeFunction(m);
+    })
+    return {...c, methods: optimizedMethods};
+}
+
+export function generateEnvironmentProgram(program: Program<any>): [Map<string, Env>, Map<string, Env>]{
+    var initialEnv = computeInitEnv(program.inits, false);
+
+    var inEnvMapping: Map<string, Env> = new Map<string, Env>();
+    var outEnvMapping: Map<string, Env> = new Map<string, Env>();
+
+    var dummyEnv = computeInitEnv(program.inits, true);
+
+    program.body.forEach(f => {
+        inEnvMapping.set(f.label, duplicateEnv(dummyEnv));
+        outEnvMapping.set(f.label, duplicateEnv(dummyEnv));
+    });
+
+    var [preds, succs, blockMapping]: [Map<string, string[]>, Map<string, string[]>, Map<string, BasicBlock<any>>] = computePredecessorSuccessor(program.body);
+
+    preds.set(program.body[0].label, ["VD"]);
+    succs.set("VD", [program.body[0].label]);
+    outEnvMapping.set("VD", initialEnv);
+
+    generateEnvironments([program.body[0].label], inEnvMapping, outEnvMapping, preds, succs, blockMapping);
+
+    return [inEnvMapping, outEnvMapping];
+}
+
+export function generateEnvironmentFunctions(func: FunDef<any>): [Map<string, Env>, Map<string, Env>]{
+    var initialEnv  = computeInitEnv(func.inits, false);
+    addParamsToEnv(func.parameters, initialEnv, false);
+
+    var inEnvMapping: Map<string, Env> = new Map<string, Env>();
+    var outEnvMapping: Map<string, Env> = new Map<string, Env>();
+
+    var dummyEnv = computeInitEnv(func.inits, true);
+    addParamsToEnv(func.parameters, initialEnv, true);
+
+    func.body.forEach(f => {
+        inEnvMapping.set(f.label, duplicateEnv(dummyEnv));
+        outEnvMapping.set(f.label, duplicateEnv(dummyEnv));
+    });
+
+    inEnvMapping.set(func.body[0].label, initialEnv);
+
+    var [preds, succs, blockMapping]: [Map<string, string[]>, Map<string, string[]>, Map<string, BasicBlock<any>>] = computePredecessorSuccessor(func.body);
+
+    preds.set(func.body[0].label, ["VD"]);
+    succs.set("VD", [func.body[0].label]);
+    outEnvMapping.set("VD", initialEnv);
+    
+    generateEnvironments([func.body[0].label], inEnvMapping, outEnvMapping, preds, succs, blockMapping);
+
+    return [inEnvMapping, outEnvMapping];
+}
+
+export function optimizeProgram(program: Program<any>): Program<any>{
```

Comment:
Generally the structure of this all makes sense to me, and follows patterns we set up in other places. Thanks for your attention to that!